guides/source/security.md

```diff
@@ -29,6 +30,147 @@ The threats against web applications include user account hijacking, bypass of a
+
+Here is a list of modified files:
+
+```bash
+On branch main
+Changes not staged for commit:
+  (use "git add <file>..." to update what will be committed)
+  (use "git restore <file>..." to discard changes in working directory)
+  modified:   Gemfile
+  modified:   Gemfile.lock
+  modified:   app/controllers/application_controller.rb
+  modified:   config/routes.rb
+
+Untracked files:
+  (use "git add <file>..." to include in what will be committed)
+  app/controllers/concerns/authentication.rb
+  app/controllers/passwords_controller.rb
+  app/controllers/sessions_controller.rb
+  app/mailers/passwords_mailer.rb
+  app/models/current.rb
+  app/models/session.rb
+  app/models/user.rb
+  app/views/passwords/
+  app/views/passwords_mailer/
+  app/views/sessions/
+  db/migrate/
+  db/schema.rb
+  test/mailers/previews/
+```
+
+### Reset Password
+
+The authentication generator also adds reset password functionality. You can see a "forgot password?" link on the "sign in" page. Clicking that link navigates to the `/passwords/new` path and routes to the passwords controller. The `new` method of the `PasswordsController` class runs through the flow for sending a password reset email.
+
+The mailers for *reset password* are also set up by the generator at `app/mailers/password_mailer.rb` and render the following email to send to the user:
+
+```html+erb
+# app/views/passwords_mailer/reset.html.erb
+<p>
+  You can reset your password within the next 15 minutes on
+  <%= link_to "this password reset page", edit_password_url(@user.password_reset_token) %>.
+</p>
+```
+
+### Implementation Details
+
+This section covers some of the implementation details for the authentication generator in Rails: The `has_secure_password` method, the `authenticate_by` method, and the `Authentication` concern.
+
+#### `has_secure_password`
+
+The [`has_secure_password`](https://api.rubyonrails.org/classes/ActiveModel/SecurePassword/ClassMethods.html#method-i-has_secure_password) method is added to the `user` model and takes care of storing a hashed password using the `bcrypt` algorithm:
+
+```ruby
+class User < ApplicationRecord
+  has_secure_password
+  has_many :sessions, dependent: :destroy
+
+  normalizes :email_address, with: -> e { e.strip.downcase }
+end
+```
+
+#### `authenticate_by`
+
+The [`authenticate_by`](https://api.rubyonrails.org/classes/ActiveRecord/SecurePassword/ClassMethods.html) method is used in the `SessionsController` while creating a new session to validate that the credentials provided by the user match the credentials stored in the database (e.g. password):
+
+```ruby
+class SessionsController < ApplicationController
+  def create
+    if user = User.authenticate_by(params.permit(:email_address, :password))
```

Comment:
nit: I would do

```ruby
def authentication_params
   params.permit(:email_address, :password)
end
```

Then do

user = User.authenticate_by(authentication_params)
Just so that if someone copy and pastes this example, its more in line with what we would see in a production app.

Reply:
@Josiassejod1 this isn't what the authentication generator generates though. If you look at the [generator template](https://github.com/rails/rails/blob/main/railties/lib/rails/generators/rails/authentication/templates/app/controllers/sessions_controller.rb.tt) for the `SessionsController` you can see that it uses the syntax:

```ruby
if user = User.authenticate_by(params.permit(:email_address, :password))
```
So IMO it makes sense to show that in this guide since the guide is discussing the code created by the generator.

Reply:
That's fair ! Thanks for the clarity and explanation.

Reply:
Yes @Josiassejod1 that is a good observation :) But we're sticking with the code that's directly from the generator template here.